### ISSUE TITLE
Make rabbitmqctl operations support a timeout

### ISSUE THREAD

**Issue by michaelklishin** [rabbitmq/rabbitmq-server#62]
The smallest in scope solution was submitted as #61 to be considered for `3.5.0`. The team has decided that it is not really an improvement over [timeout(1)](http://linux.die.net/man/1/timeout) (except for working on Windows, perhaps) and ideally the `list_*` commands should list unresponsive/unreachable
items (e.g. channels or queues).

[Mailing list thread](https://groups.google.com/forum/#!searchin/rabbitmq-users/hangs/rabbitmq-users/Ja0iyfF0Szw/MHm8RuqCaacJ).

**Comment by michaelklishin** [comment]
Given ongoing community requests for this feature, we will prioritise this. #61 may be resurrected and included into `3.5.x`.

**Comment by michaelklishin** [comment]
First improvements landed in #181, will be in `3.5.x`.

**Comment by dumbbell** [comment]
We discussed this internaly some time ago, here are more details.

To go beyond what was committed in #181, we want to handle timeous in a finer grain way. Let's take `list_queues` as an example: if there are 10 queues, but one of them seems stuck, the current implementation will timeout globally and may display no information about the responsive queues.

An improved implementation would display information about responsive queues, and indicate a timeout for each stuck element.

**Comment by videlalvaro** [comment]
Perhaps the "one second timeout" could be specified either per element or globally, the later being the default one

**Comment by michaelklishin** [comment]
Yes, this is a good question whether we should have per-element timeouts. I don't think this is what our users would expect. "Global" timeouts are much more common in similar tools.

**Comment by michaelklishin** [comment]
Also note that as far as ops engineers are concerned, it's the total timeout that they care about most.

**Comment by dumbbell** [comment]
I'm not sure a per-element timeout is interesting. Furthermore it would add another option and complicate the UI.

**Comment by videlalvaro** [comment]
I think it's fine as long as we output things as they happen. ie: we print as much stuff as we can during the specified timeout; if the timeout happens, then fine. Stop the whole thing

**Comment by dumbbell** [comment]
It also depends if the output must be sorted or not. I admit I didn't verify if it is the case already for eg. `list_queues`. If we don't need to sort the output, I agree we should display everything as soon as we get the information.

**Comment by Ayanda-D** [comment]
The output is sorted. See `rabbit_control_main:display_info_list`. Question is, do we keep it as is, and display output only after we've received all info and sorted it, or, remove the sorting and display things as they happen?

**Comment by michaelklishin** [comment]
@Ayanda-D relying in rabbitmqctl output to be sorted is wrong, so if anybody does that, that's too bad. Don't worry about ordering. We've never gotten any feedback on the sorting behaviour.

**Comment by Ayanda-D** [comment]
Okay thanks @michaelklishin . Below is the current solution proposal and implementation steps I'll be following. Idea is to make responses from the rabbit nodes independent of the direct rpc:call's from `rabbit_cli.erl` module, allowing things to be displayed as they happen.

1. `rabbitmqctl` command operates by starting a temporary node upon execution, from which remote calls to rabbit nodes are executed from.
2. The `rabbit_cli:rpc_call` will be a spawned process, separate from the process running on the shell.
3. The user defined timeout will be passed as part of the rpc:call, as already done in [#181](https://github.com/rabbitmq/rabbitmq-server/pull/181). In addition to this timeout being passed, the PID of the calling process i.e. self(), will also be included as part of the rpc:call params.
4. After spawning `rabbit_cli:rpc_call`, the calling process will enter a receive loop, with a timeout equal to the user defined timeout, and wait to display responses it receives.
5. At the backend, most operations make use of list comprehensions during data acquisition from mnesia for example. These will be updated to send message replies to the calling process upon execution of each element/step.
6. The calling process will come out of the receive loop when it receives a `finished` message or when it times out. After this, program flow proceeds and the temporary node terminates.
7. Any delayed messages from the rabbit nodes will be handled as in the current implementation; unhandled basically. They'll be sent to a nonexistant process which would have terminated at the end of the `rabbitmqctl` operation in (6), or be in the message queue of the spawned process in (2) if the temporary node hasn't terminated.

There was also concern about delayed messages sent to the process doing the `rpc:call`, from operations such as `list_queues` and `list_connections`, which do gen_server calls behind the scenes with timeouts set to `infinity`. However these won't be of concern as mentioned in (7).

**Comment by michaelklishin** [comment]
@Ayanda-D sounds good to me.

**Comment by Ayanda-D** [comment]
@michaelklishin okay great. I'll proceed with the implementation.

**Comment by michaelklishin** [comment]
Changing this to `effort-medium` because this turned out to be much more time consuming in practice.

**Comment by michaelklishin** [comment]
Also relevant: #396.
